Feature: inviting friends to events

Eukon05 opened the pull request:
- Added the ability to invite friends to events
- Fixed an error when a user would remain in "invitees" after confirming their attendance

Eukon05 committed:
- Added the ability to invite friends to events
- Fixed an error when a user would remain in "invitees" after confirming their attendance

krzysztofcybulski commented on src/test/java/pl/eukon05/eventboard/event/application/service/UnitTests.java:
```diff
@@ -18,12 +20,14 @@
     private final ModifyEventUseCase modifyEventUseCase = new ModifyEventUseCase(saveEventPort, getEventPort);
     private final DeleteEventUseCase deleteEventUseCase = new DeleteEventUseCase(getEventPort, deleteEventPort);
     private final AttendEventUseCase attendEventUseCase = new AttendEventUseCase(getEventPort, saveEventPort);
+    private final InviteToEventUseCase inviteToEventUseCase = new InviteToEventUseCase(getUserFriendlistPort, getEventPort, saveEventPort);
```
NTH: Maybe split test classes per tested use case? Will be easier to maintain

Eukon05 replied:
Good idea, these classes got really cluttered very fast

krzysztofcybulski commented on src/test/java/pl/eukon05/eventboard/event/application/service/UnitTests.java:
```diff
@@ -140,4 +144,38 @@ void should_attend_invited_event() {
+    }
+
+    @Test
+    void should_invite_to_private_event() {
```
What do you think about testing some negative paths?

krzysztofcybulski commented on src/test/java/pl/eukon05/eventboard/event/application/service/UnitTests.java:
```diff
@@ -140,4 +144,38 @@ void should_attend_invited_event() {
+
+    @Test
+    void should_invite_to_private_event() {
+        Event event = createTestEvent();
```
```suggestion
        Event event = createTestPrivateEvent();
```

krzysztofcybulski commented on src/test/java/pl/eukon05/eventboard/event/application/service/UnitTests.java:
```diff
@@ -140,4 +144,38 @@ void should_attend_invited_event() {
+        event.setOrganizerID(userID);
+
+        String friendID = "somerandomid";
+        Mockito.when(getEventPort.getEventById(Mockito.anyLong())).thenReturn(Optional.of(event));
```
NTH: Maybe you can think about extracting complex lines?
```suggestion
        gettingEventByIdWillReturn(event);
```

krzysztofcybulski commented on src/test/java/pl/eukon05/eventboard/event/application/service/UnitTests.java:
```diff
@@ -140,4 +144,38 @@ void should_attend_invited_event() {
+
+        String friendID = "somerandomid";
+        Mockito.when(getEventPort.getEventById(Mockito.anyLong())).thenReturn(Optional.of(event));
+        Mockito.when(getUserFriendlistPort.getUserFriendlist(Mockito.anyString())).thenReturn(List.of(friendID));
```
```suggestion
        gettingUserFriendlistWillReturn(friendId)
```

krzysztofcybulski commented on src/test/java/pl/eukon05/eventboard/event/application/service/UnitTests.java:
```diff
@@ -140,4 +144,38 @@ void should_attend_invited_event() {
+        Mockito.when(getEventPort.getEventById(Mockito.anyLong())).thenReturn(Optional.of(event));
+        Mockito.when(getUserFriendlistPort.getUserFriendlist(Mockito.anyString())).thenReturn(List.of(friendID));
+
+        assertTrue(inviteToEventUseCase.execute(userID, friendID, event.getId()));
```
I would move `when` section (`inviteToEventUseCase.execute`) out of `then`.
```suggestion
        var invitiationResult = inviteToEventUseCase.execute(userID, friendID, event.getId());
        
        assertTrue(invitiationResult)
```

krzysztofcybulski commented on src/test/java/pl/eukon05/eventboard/event/application/service/UnitTests.java:
```diff
@@ -140,4 +144,38 @@ void should_attend_invited_event() {
+
+        assertTrue(inviteToEventUseCase.execute(userID, friendID, event.getId()));
+
+        event.setOrganizerID("");
```
Not sure what happens here?

Eukon05 replied:
Well, I did something stupid and basically combined two different test cases into a single test... I'll move this code to a separate method

krzysztofcybulski commented on src/test/java/pl/eukon05/eventboard/event/application/service/UnitTests.java:
```diff
@@ -140,4 +144,38 @@ void should_attend_invited_event() {
+        Mockito.when(getUserFriendlistPort.getUserFriendlist(Mockito.anyString())).thenReturn(List.of(friendID));
+
+        assertTrue(inviteToEventUseCase.execute(userID, friendID, event.getId()));
+        assertTrue(event.getInviteeIDs().contains(friendID));
```
Doesn't JUnit have better assertions than `assertTrue`? Maybe you can use `Hamcrest` or write your own assertions?

I would love to see something like:
```suggestion
        assertInviriteeIDsContains(event, friendID);
```

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/application/service/InviteToEventUseCase.java:
```diff
@@ -0,0 +1,43 @@
+    private final GetEventPort getEventPort;
+    private final SaveEventPort saveEventPort;
+
+    boolean execute(String selfID, String friendID, long eventID) {
```
What does `true/false` mean here? :)

Eukon05 replied:
It's a placeholder for an object that will tell if the operation was successful, and if not, contain a message detailing what went wrong. I've decided to use true (success) / false (failure) for now, but I'll change it soon.

krzysztofcybulski replied:
Thanks, I like this is a conscious decision 👍🏻 Normally I would go with simple interface `Result` with two children `Success` and `Failure` from the start

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/application/service/InviteToEventUseCase.java:
```diff
@@ -0,0 +1,43 @@
+    private final GetEventPort getEventPort;
+    private final SaveEventPort saveEventPort;
+
+    boolean execute(String selfID, String friendID, long eventID) {
```
Consider packing ids into value classes. (example article: https://betterprogramming.pub/why-you-should-avoid-using-primitive-types-cb55857baa39)

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/adapter/in/domain/FriendlistAdapter.java:
```diff
@@ -0,0 +1,28 @@
+package pl.eukon05.eventboard.event.adapter.in.domain;
```
I don't get the package structure. `domain` inside of `adapter`? 🤔

Eukon05 replied:
I've named the package "domain", because it's for adapters responsible for fetching data from another domains...
Maybe I should make packages for every domain? For example, I could rename this package to "User", because the adapter classes in it connect to the User domain?
The package structure would look like this:
adapter -> in (because it's for loading external data) -> user (because it connects to the User domain)
Do you think that would make sense?

krzysztofcybulski replied:
I see 🤔 Now it starts to make some sense. If this package structure is useful for you, stay with that. 
I like to use more flat structure, so e.g. `adapter.friends.FriendlistAdapter`, `adapter.x.Y` etc.

Also I don't feel that this is an `in` adapter. We are calling it, and it's going `out` of our domain. I would classify e.g. REST API as `in`, because it calls our domain (from outside)

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/adapter/in/domain/FriendlistAdapter.java:
```diff
@@ -0,0 +1,28 @@
+
+        Optional<User> user = getUserPort.getUserById(userID);
+        if (user.isPresent())
+            friendlist = user.get().getFriends().stream().map(User::getId).toList();
```
Avoid using `get()` on `Optional`. It has much richer api.
```suggestion
            @Override
    public List<String> getUserFriendlist(String userID) {
        return getUserPort
                .getUserById(userID)
                .map(this::friendIds)
                .orElse(emptyList());
    }

    private List<String> friendIds(User user) {
        return user.getFriends().stream().map(User::getId).toList();
    }
```

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/application/port/in/GetUserFriendlistPort.java:
```diff
@@ -0,0 +1,7 @@
+import java.util.List;
+
+public interface GetUserFriendlistPort {
+    List<String> getUserFriendlist(String userID);
```
```suggestion
    List<Friend> getUserFriendlist(UserId userID);
```

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/application/service/InviteToEventUseCase.java:
```diff
@@ -0,0 +1,43 @@
+        if (event.getType().equals(EventType.PRIVATE) && !event.getGuestIDs().contains(selfID) && !event.getOrganizerID().equals(selfID))
+            return false;
+
+        event.getInviteeIDs().add(friendID);
```
[Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter). This is general comment, to the whole code.


```suggestion
        event.inite(friendId);
```

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/application/service/InviteToEventUseCase.java:
```diff
@@ -0,0 +1,43 @@
+        if (event.getGuestIDs().contains(friendID) || event.getInviteeIDs().contains(friendID)) return false;
+
+        if (event.getType().equals(EventType.PRIVATE) && !event.getGuestIDs().contains(selfID) && !event.getOrganizerID().equals(selfID))
+            return false;
```
These `return false` are very tricky :) I wonder if you don't want to show some notifications in API?

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/application/service/InviteToEventUseCase.java:
```diff
@@ -0,0 +1,43 @@
+
+        Event event = eventOptional.get();
+
+        List<String> friendlist = getUserFriendlistPort.getUserFriendlist(selfID);
```
If the port is only used to check if someone is friend of someone why do why have to return list of all friends?

```suggestion
        friendsPort.areUsersFriends(selfID, friendId);
```

Eukon05 replied:
Now that you mention it, I don't know either 😅
I'll re-do it to return a boolean value, depending on if the user in question is a "friend" or not

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/application/service/InviteToEventUseCase.java:
```diff
@@ -0,0 +1,43 @@
+
+        if (!friendlist.contains(friendID)) return false;
+
+        if (event.getGuestIDs().contains(friendID) || event.getInviteeIDs().contains(friendID)) return false;
```
Shouldn't `Event` be responsible for this?

```java
class Event {

  public void invite(String userId) {
        if(!guestIDs.contains(userId) && !inviteeIDs.contains(userId)) {
            inviteeIDs.add(userId);
        }
    }

}
```

krzysztofcybulski commented on src/main/java/pl/eukon05/eventboard/event/application/service/InviteToEventUseCase.java:
```diff
@@ -0,0 +1,43 @@
+
+        if (event.getGuestIDs().contains(friendID) || event.getInviteeIDs().contains(friendID)) return false;
+
+        if (event.getType().equals(EventType.PRIVATE) && !event.getGuestIDs().contains(selfID) && !event.getOrganizerID().equals(selfID))
```
This is also `Event` responsibility. Domain should guard consistency. Use case should only "call" domain, but not contain business logic.

Eukon05 replied:
OK, I'll move these types of tasks to dedicated methods in the domain objects.

krzysztofcybulski reviewed (changes requested):
Thanks for sharing the code! ;) I tried to give you some advices, if you don't agree with something or need more explanation let me know.

I really like the code, hope that my comments will make it even better 💯

Eukon05 committed:
- Renamed FriendlistAdapter to FriendsAdapter, streamlined its functionality and moved it to a new package, better reflecting its purpose
- Renamed GetUserFriendlistPort to CheckIfFriendsPort, now it returns a boolean value, indicating if two users are friends, instead of returning a user's friendlist
- Moved the guard clauses regarding inviting users to events to the Event domain object
- Moved the guard clauses regarding confirming attendance to the Event domain object
- Split test classes for both Event and User domains into multiple classes for better maintainability
- Replaced usage of `assertTrue(somecollection.contains(someobject))` with `assertThat(somecollection, contains(someobject))` from Hamcrest

Massive thanks to @krzysztofcybulski for his feedback!

Eukon05 commented:
Thank you for your feedback!
I've made some changes according to your comments and I'd be very glad if you could take a look a them.
There are some things I will correct in future commits, such as missing negative paths in tests, but I hope that you'll find the current changes appropriate.

Eukon05 committed:
- Added two negative path tests to InviteToEventUnitTests
- Replaced long lines in InviteToEventUnitTests with method calls

Eukon05 committed:
- Added a test for inviting a friend to a private event, that is NOT hosted by the inviter
